client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1035 @@
+        }
+        kaa_list_destroy(self->optional_listeners, &kaa_data_destroy);
+    }
+    if (self->uids) {
+        kaa_list_destroy(self->uids, &kaa_uids_destroy);
+    }
+    if (self->topics) {
+        kaa_list_destroy(self->topics, &kaa_topics_destroy);
+    }
+
+    KAA_FREE(self);
+}
+
+kaa_error_t kaa_add_notification_listener(kaa_notification_manager_t *self, kaa_notification_listener_t *listener, uint32_t* listener_id)
+{
+    KAA_RETURN_IF_NIL2(self, listener, KAA_ERR_BADPARAM);
+    if (!listener->callback) {
+        KAA_LOG_ERROR(self->logger, KAA_ERR_BADPARAM, "Expected callback.Got NULL.");
+        return KAA_ERR_BADPARAM;
+    }
+    uint32_t id;
+    kaa_error_t err = kaa_calculate_notification_listener_id(listener, &id);
+    if (err) {
+        KAA_LOG_WARN(self->logger, err, "Failed to calculate mandatory listener id.");
+        return err;
+    }
+    if (kaa_list_find_next(self->mandatory_listeners, &kaa_find_notification_listener_by_id, &id)) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_ALREADY_EXISTS, "Failed to add mandatory listener: listener is already subscribed.");
+        return KAA_ERR_ALREADY_EXISTS;
+    }
+    kaa_notification_listener_wrapper_t* wrapper = (kaa_notification_listener_wrapper_t *) KAA_MALLOC(sizeof(kaa_notification_listener_wrapper_t));
+    KAA_RETURN_IF_NIL(wrapper, KAA_ERR_NOMEM);
+    wrapper->listener = *listener;
+    kaa_list_t *new_listener = NULL;
+    if (!self->mandatory_listeners) {
+        new_listener = kaa_list_create(wrapper);
+    } else {
+        new_listener = kaa_list_push_front(self->mandatory_listeners, wrapper);
+    }
+    if (!new_listener) {
+        KAA_FREE(wrapper);
+        KAA_LOG_WARN(self->logger, KAA_ERR_ALREADY_EXISTS, "Failed to add mandatory listener.");
+        return KAA_ERR_NOMEM;
+    }
+    self->mandatory_listeners = new_listener;
+    *listener_id = wrapper->id = id;; // for user to have convenient way to address notification listener*/
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_remove_notification_listener(kaa_notification_manager_t *self, uint32_t *listener_id)
+{
+    KAA_RETURN_IF_NIL2(self, listener_id, KAA_ERR_BADPARAM);
+    kaa_error_t err = kaa_list_remove_first(&self->mandatory_listeners, &kaa_find_notification_listener_by_id, listener_id, &kaa_data_destroy);
+    if (err) {
+        KAA_LOG_WARN(self->logger, err, "Failed to remove mandatory listener: listener with id = %lu is not found.", *listener_id);
+    }
+    if (!kaa_list_get_size(self->mandatory_listeners)) {
+        KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Mandatory listeners list is empty now.");
+    }
+    return err;
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_add_optional_notification_listener(kaa_notification_manager_t *self, kaa_notification_listener_t *listener
+                                                 , uint64_t *topic_id, uint32_t *listener_id)
+{
+    KAA_RETURN_IF_NIL4(self, listener, topic_id, listener_id, KAA_ERR_BADPARAM);
```

Comment:
Allow a user to pass NULL instead of listener_id. Do corresponding changes in all function to avoid dereferencing NULL pointer.